packages/flutter/lib/src/widgets/sheet_route.dart

```diff
@@ -0,0 +1,266 @@
+import 'dart:ui';
+import 'package:flutter/foundation.dart';
+import 'package:flutter/widgets.dart';
+
+import 'sheet.dart';
+
+/// Route that defines the transition animation of the previous route when this
+/// one is closing
+mixin RouteWithPreviousTransitionMixin<T> on Route<T> {
+  Widget getPreviousRouteTransition(
+    BuildContext context,
+    Animation<double> secondAnimation,
+    Widget child,
+  );
+}
+
+/// Route that allows the next route to define the animation transition of this route
+/// when the route appears back after the next one is popped
+mixin AllowsRouteWithPreviousTransitionMixin<T> on PageRoute<T> {
+  RouteWithPreviousTransitionMixin? _nextModalRoute;
+
+  @override
+  bool canTransitionTo(TransitionRoute<dynamic> nextRoute) {
+    return super.canTransitionTo(nextRoute) ||
+        (nextRoute is RouteWithPreviousTransitionMixin);
+  }
+
+  @override
+  void didChangeNext(Route? nextRoute) {
+    if (nextRoute is RouteWithPreviousTransitionMixin) {
+      _nextModalRoute = nextRoute;
+    }
+
+    super.didChangeNext(nextRoute);
+  }
+
+  @override
+  bool didPop(T result) {
+    _nextModalRoute = null;
+    return super.didPop(result);
+  }
+
+  @override
+  Widget buildTransitions(
+    BuildContext context,
+    Animation<double> animation,
+    Animation<double> secondaryAnimation,
+    Widget child,
+  ) {
+    if (_nextModalRoute != null) {
+      if (secondaryAnimation.isDismissed) {
+        _nextModalRoute = null;
+      } else {
+        // Avoid default transition theme to animate when a new modal view is pushed
+        final Animation<double> fakeSecondaryAnimation =
+            Tween<double>(begin: 0, end: 0).animate(secondaryAnimation);
+
+        final Widget defaultTransition = super.buildTransitions(
+            context, animation, fakeSecondaryAnimation, child);
+        return _nextModalRoute!.getPreviousRouteTransition(
+            context, secondaryAnimation, defaultTransition);
+      }
+    }
+    return super.buildTransitions(
+      context,
+      animation,
+      secondaryAnimation,
+      child,
+    );
+  }
+}
+
+const Duration _bottomSheetDuration = Duration(milliseconds: 400);
+
+class _Sheet<T> extends StatefulWidget {
```

Comment:
I'm a bit surprised there's a _Sheet and _SheetState here and another Sheet and _SheetState in the other file. Maybe the names need to be more specific about what they do.